velox/exec/GroupingSet.cpp

```diff
@@ -1081,6 +1094,48 @@ bool GroupingSet::mergeNext(
   VELOX_UNREACHABLE();
 }
 
+bool GroupingSet::mergeNextForDistinct(
+    int32_t maxOutputRows,
+    int32_t maxOutputBytes,
+    const RowVectorPtr& result) {
+  VELOX_CHECK(isDistinct());
+
+  bool nextKeyIsEqual{false};
+  bool newDistinct{true};
+  int32_t numOutputRows{0};
+  for (;;) {
+    const auto next = merge_->nextWithEquals();
+    auto* stream = next.first;
+    if (stream == nullptr) {
+      result->resize(numOutputRows);
+      break;
+    }
+    if (stream->id() == 0) {
+      // A spilled row is not distinct if it is contained in the first stream
+      // which has already been output as distinct before we triggered the
+      // spilling.
+      newDistinct = false;
+    }
+    nextKeyIsEqual = next.second;
+    if (!nextKeyIsEqual) {
```

Comment:
I find it difficult to follow this logic. I wonder if there is a way to more readable way to write this.

```
if (next.second) {
   // Skip duplicate rows until we get to the last one.
  stream->pop();
  continue;
}

// Check if we already produced this row before spilling started.
if (stream->id() == 0) {
  stream->pop();
  continue;
}

// Yield result.
result->copy(&stream->current(), numOutputRows++, stream->currentIndex(), 1);
stream->pop();

if (numOutputRows >= maxOutputRows) {
   break;
}
```

Reply:
I make a change to simplify the loop. Thanks!